tools/syz-declextract/syz-declextract.cpp

```diff
@@ -162,41 +171,163 @@ class Matcher : public MatchFinder::MatchCallback {
-      // The array could have an implicitly initialized policy (i.e. empty)
-      if (policy->children().empty()) {
-        continue;
-      }
+
+    const auto *init = netlinkDecl->getInit();
+    if (!init) {
+      return;
+    }
+    for (const auto &policy : *llvm::dyn_cast<InitListExpr>(init)) {
       fields.push_back(std::vector<Expr *>());
       for (const auto &member : policy->children()) {
         fields.back().push_back(llvm::dyn_cast<Expr>(member));
       }
     }
 
-    const auto enumData = enumMatcher.getEnumData();
+    EnumMatcher enumMatcher(fields.size());
+    MatchFinder enumFinder;
+    enumFinder.addMatcher(decl(forEachDescendant(designatedInitExpr(
+                              optionally(has(constantExpr(has(declRefExpr())).bind("enum_value")))))),
+                          &enumMatcher);
+    enumFinder.match(*netlinkDecl, *context); // get enum details from the current subtree (nla_policy[])
+    auto enumData = enumMatcher.getEnumData();
+
     for (const auto &item : enumData) {
+      if (item.file.empty()) {
+        continue;
+      }
+      if (item.file.back() != 'h') { // only extract from "*.h" files
+        return;
+      }
       printf("include<%s>\n", item.file.c_str());
     }
-    printf("%s$auto[\n", netlinkDecl->getDefinition()->getNameAsString().c_str());
+
+    printf("%s[\n", getPolicyName(Result, netlinkDecl)->c_str());
     for (size_t i = 0; i < fields.size(); ++i) {
-      printf("\t%s nlattr[%s, %s]\n", enumData[i].name.c_str(), enumData[i].name.c_str(),
-             nlaToSyz(fields[i][0]).c_str());
+      // The array could have an implicitly initialized policy (i.e. empty)
+      if (fields[i].empty()) {
+        continue;
+      }
+      if (enumData[i].name.empty()) {
+        printf("\tUNNAMED_ATTR_%zu nlattr[%zu, %s]\n", i, i, nlaToSyz(fields[i][0]).c_str());
+      } else {
+        printf("\t%s nlattr[%s, %s]\n", enumData[i].name.c_str(), enumData[i].name.c_str(),
+               nlaToSyz(fields[i][0]).c_str());
+      }
     }
     puts("] [varlen]");
   }
 
+  std::map<std::string, unsigned> genlFamilyMember;
+  std::map<std::string, std::map<std::string, unsigned>> opsMember;
+
+  std::optional<std::string> getPolicyName(const MatchFinder::MatchResult &Result, const ValueDecl *decl) {
+    if (!decl) {
+      return std::nullopt;
+    }
+    std::string filename =
+        std::filesystem::path(
+            Result.SourceManager->getFilename(decl->getCanonicalDecl()->getSourceRange().getBegin()).str())
+            .filename()
+            .stem()
+            .string();
+    std::replace(filename.begin(), filename.end(), '-', '_');
+    return std::make_optional(decl->getNameAsString() + "$auto_" + filename);
+  }
+
+  std::vector<Ops> getOps(const MatchFinder::MatchResult &Result, const std::string opsName, const InitListExpr *init) {
+    ASTContext *context = Result.Context;
+    std::vector<Ops> cmds;
```

Comment:
This should be called `ops` since all other code refers to this as "ops" rather than "cmds".

Reply:
Done!